sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -339,8 +339,11 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
       case SqlBaseParser.SELECT =>
         // Regular select
 
+        // Add hints.
+        val withHint = relation.optionalMap(ctx.hint)(withHints)
```

Comment:
for now the hint can only affect the relations(broadcast hint), so it's ok to just put the `Hint` on top of relation. But generally hint can affect anything in the select clause, e.g. maybe the filter condition.